source/dub/generators/generator.d

```diff
@@ -143,130 +142,151 @@ class ProjectGenerator
-			else if (tt == TargetType.dynamicLibrary) {
-				logWarn("Dynamic libraries are not yet supported as dependencies - building as static library.");
-				tt = TargetType.staticLibrary;
-			}
-		}
-		if (tt != TargetType.none && tt != TargetType.sourceLibrary && shallowbs.sourceFiles.empty) {
-			logWarn(`Configuration '%s' of package %s contains no source files. Please add {"targetType": "none"} to its package description to avoid building it.`,
-				configs[pack.name], pack.name);
-			tt = TargetType.none;
-		}
-
-		shallowbs.targetType = tt;
-		bool generates_binary = tt != TargetType.sourceLibrary && tt != TargetType.none;
-		bool is_target = generates_binary || pack is m_project.rootPackage;
+		BuildSettings buildsettings, forDependenciesBuildSettings;
+		bool is_target;
 
-		if (tt == TargetType.none) {
-			// ignore any build settings for targetType none (only dependencies will be processed)
-			shallowbs = BuildSettings.init;
-			shallowbs.targetType = TargetType.none;
+		void setIs_target(TargetType tt)
+		{
+			bool generates_binary = tt != TargetType.sourceLibrary && tt != TargetType.none;
+			is_target = generates_binary || pack is m_project.rootPackage;
 		}
 
-		// start to build up the build settings
-		BuildSettings buildsettings;
-		processVars(buildsettings, m_project, pack, shallowbs, true);
+		static auto mergeFromDependents(BuildSettings parent, ref BuildSettings child) {
+			child.addVersions(parent.versions);
+			child.addDebugVersions(parent.debugVersions);
+			child.addOptions(BuildOptions(cast(BuildOptions)parent.options & inheritedBuildOptions));
+
+			// special support for overriding string imports in parent packages
+			// this is a candidate for deprecation, once an alternative approach
+			// has been found
+			if (child.stringImportPaths.length) {
+				// override string import files (used for up to date checking)
+				foreach (ref f; child.stringImportFiles)
+					foreach (fi; parent.stringImportFiles)
+						if (f != fi && Path(f).head == Path(fi).head) {
+							f = fi;
+						}
 
-		// remove any mainSourceFile from library builds
-		if (buildsettings.targetType != TargetType.executable && buildsettings.mainSourceFile.length) {
-			buildsettings.sourceFiles = buildsettings.sourceFiles.filter!(f => f != buildsettings.mainSourceFile)().array;
-			main_files ~= buildsettings.mainSourceFile;
+				// add the string import paths (used by the compiler to find the overridden files)
+				child.prependStringImportPaths(parent.stringImportPaths);
+			}
+
+			return child;
 		}
 
-		// set pic for dynamic library builds.
-		if (buildsettings.targetType == TargetType.dynamicLibrary)
-			buildsettings.addOptions(BuildOption.pic);
+		auto packInTargets = pack.name in targets;
+		if (packInTargets is null) {
+			// determine the actual target type
+			auto shallowbs = pack.getBuildSettings(settings.platform, configs[pack.name]);
```

Comment:
This is the raw build settings as defined in the package recipe, filtered for the actual platform/config. It doesn't include settings inherited from other packages or settings based on the current build mode.

Reply:
Kind of guessed that, there is also [BuildSettingsTemplate](https://github.com/dlang/dub/blob/255209da6491f98df602b121a8effe2c1fc94e69/source/dub/recipe/packagerecipe.d#L136).
Somewhat confusing to use that term both in the receipt and in the compiler domain.

Would (Abstract/Generic)CompilerArgs or CompilerSettings be a more appropriate name?
The term Target is also fairly well know in build tools and seems to be a good fit.

I'm stopping here as this is off-topic, but unclear naming is a major hurdle for me to work on dub or review complex PRs.

Reply:
I'd say not really, because it really does include all build settings, not just compiler specific ones. But `BuildSettingsRecipe` would actually be a better fit than "template" (the term "recipe" was introduced later but arguably does a much better job explaining what it is).